src/EntityComponentManager.cc

```diff
@@ -963,6 +963,37 @@ std::unordered_set<ComponentTypeId>
   return periodicComponents;
 }
 
+/////////////////////////////////////////////////
+void EntityComponentManager::PeriodicChangeEntityComponentMap(
+  std::unordered_map<Entity,
+    std::unordered_set<ComponentTypeId>> &_changes) const
+{
+  // Get all changes
+  for (const auto &[componentType, entities] :
+    this->dataPtr->periodicChangedComponents)
+  {
+    for (const auto entity : entities)
+    {
+      _changes[entity].emplace(componentType);
+    }
+  }
+
+  // Get all removed components
+  for (const auto &[entity, components] :
```

Comment:
Hmm, I put this here because I was not sure if it makes sense to add a changed state component when a component has been removed to the serialized message. The reason being the `msg::SerializedStateMap` may have conflicting instructions to remove and/or modify the state. It may make sense to move it to the `PeriodicStateFromCache` function so we aren't checking for removed objects over and over again.

Reply:
I looked into what `SceneBroadcaster` does when a component is removed. It uses `ChangedState` instead of sending the full state, which means some periodic changes might be missed, so we need to keep the cache (not clear it) after a component removal. Your code already does that, so no change needed.  This also means we have to remove components from the cache if they've been removed from the ECM. So yeah, I agree we should keep the code. Moving it to `PeriodicStateFromCache` would improve performance.

Reply:
So I was tempted to move this out to the `PeriodicStateFromCache` function but it turns out that we only run that every few iterations, but the ECM invalidates `toBeRemoved` every iteration, so moving this logic there would require more work than is worth. We would also need a data structure to track an accumulation of the removed entities and components which is not ideal.